### ISSUE TITLE
xgboost model predicts based on row indices of input data set when trained using gpu

### ISSUE THREAD

**Issue by errorplagued** [dmlc/xgboost#2764]
Something strange happened when I was training an xgboost binary classification model using the gpu. After fitting on the training data set, it would make positive predictions based on the row indices of the input data set. So let's say making positive predictions on rows 1, 3, and 5 on the training set gives perfect precision (i.e. all true positives). If i feed in any other data set, the model would always make positive predictions on rows 1, 3, and 5. Below is a minimal working example to replicate the problem using the iris data set (multi-classification):

```
from xgboost import XGBClassifier
import sklearn.datasets
import numpy as np

kwargs = {'tree_method': 'gpu_exact'}
model = XGBClassifier(**kwargs)
X = sklearn.datasets.load_iris().data
Y = sklearn.datasets.load_iris().target
model.fit(X, Y)

#Predictions using the training data as input matrix
print('Training Data')
pred = model.predict(X)
p = np.array([round(x) for x in pred])
print('Predicted true positive indices (veriscolour): ' + str([i for i, x in enumerate((p==Y) & (p==1)) if x]))
print('Precision (veriscolour): ' + str(sum((p==1) & (p==Y))/float(sum(Y==1))))

#Now predict on an input zero-matrix
print('Zero-Matrix')
pred = model.predict(np.zeros(np.shape(X)))
p = np.array([round(x) for x in pred])
print('Predicted true positive indices (veriscolour): ' + str([i for i, x in enumerate((p==Y) & (p==1)) if x]))
print('Precision (veriscolour): ' + str(sum((p==1) & (p==Y))/float(sum(Y==1))))

#Now predict on the shuffled training data set
print('Shuffled training dataset')
indices = range(len(Y))
np.random.shuffle(indices)
X = X[indices, :]
Y = Y[indices]
pred = model.predict(X)
p = np.array([round(x) for x in pred])
print('Predicted true positive indices (veriscolour): ' + str([i for i, x in enumerate((p==Y) & (p==1)) if x]))
print('Precision (veriscolour): ' + str(sum((p==1) & (p==Y))/float(sum(Y==1))))
```

As you can see from the output, the indices of the predicted positives (veriscolour) in the first two data sets are the same, while the shuffled training data set returns a different (smaller) set of positives. However, if I change it to the 'exact' method instead of the 'gpu-exact' method, the problem seems to resolve. I am pretty certain I have installed CUDA correctly so I am wondering if my version of xgboost is out of date or if this is a known issue (I can't seem to find any information on this)? I installed xgboost with pip in a conda environment. Version is 0.6 and python version is 2.7.13.

**Comment by RAMitchell** [comment]
May be related to #2759

**Comment by RAMitchell** [comment]
I am unable to reproduce this on my machine. Are you sure you have installed the XGBoost GPU version correctly? You will need to build from source you should not be able to install from pip.
